src/frequenz/sdk/timeseries/battery_pool/_metric_calculator.py

```diff
@@ -618,32 +520,95 @@ def calculate(
+        battery_sets = {
+            self._bat_bats_map[battery_id] for battery_id in working_batteries
+        }
+
+        def get_validated_bounds(
+            comp_id: int, comp_metric_ids: list[ComponentMetricId]
+        ) -> PowerBounds | None:
+            results: list[float] = []
+            # Make timestamp accessible
+            nonlocal loop_timestamp
+            local_timestamp = loop_timestamp
+
+            if data := metrics_data.get(comp_id):
+                for comp_metric_id in comp_metric_ids:
+                    val = data.get(comp_metric_id)
+                    if val is not None:
+                        local_timestamp = max(loop_timestamp, data.timestamp)
+                        results.append(val)
+
+            if len(results) != len(comp_metric_ids):
+                return None
+
+            loop_timestamp = local_timestamp
+            return PowerBounds(
+                inclusion_lower=results[0],
+                exclusion_lower=results[1],
+                exclusion_upper=results[2],
+                inclusion_upper=results[3],
+            )
+
+        for battery_ids in battery_sets:
+            loop_timestamp = timestamp
+
+            inverter_ids = self._bat_inv_map[next(iter(battery_ids))]
+
+            battery_bounds = list(
+                x
+                for x in map(
+                    lambda battery_id: get_validated_bounds(
+                        battery_id, self._battery_metrics
+                    ),
+                    battery_ids,
+                )
+                if x is not None
+            )
+
+            if len(battery_bounds) == 0:
+                continue
+
+            aggregated_bat_bounds = _aggregate_battery_power_bounds(battery_bounds)
+
+            inverter_bounds = list(
+                x
+                for x in map(
+                    lambda inverter_id: get_validated_bounds(
+                        inverter_id, self._inverter_metrics
+                    ),
+                    inverter_ids,
+                )
+                if x is not None
+            )
+
+            if len(inverter_bounds) == 0:
+                continue
+
+            timestamp = max(timestamp, loop_timestamp)
+
+            inclusion_bounds_lower += max(
+                aggregated_bat_bounds.inclusion_lower,
+                sum(bound.inclusion_lower for bound in inverter_bounds),
+            )
+            inclusion_bounds_upper += min(
+                aggregated_bat_bounds.inclusion_upper,
+                sum(bound.inclusion_upper for bound in inverter_bounds),
+            )
+            exclusion_bounds_lower += min(
+                aggregated_bat_bounds.exclusion_lower,
+                sum(bound.exclusion_lower for bound in inverter_bounds),
+            )
+            exclusion_bounds_upper += max(
+                aggregated_bat_bounds.exclusion_upper,
+                sum(bound.exclusion_upper for bound in inverter_bounds),
+            )
```

Comment:
This can probably done iterating through `inverter_bounds` only once

Reply:
Yes, but it's usually not more than one inverter for 99% of the cases and it makes the code more readable this way

Reply:
fair enough, resolving...